### PR TITLE
add arm64 type for bastion instance

### PR DESCRIPTION
**Pull request by tedteng**
**How to categorize this PR?**

/area ops-productivity
/kind bug
/platform aws

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes https://github.com/gardener/gardener-extension-provider-aws/issues/694

**Special notes for your reviewer**:

**Release note**:

```other operator
add 't4g.nano' and arm64 type instance supported in the bastion instance
```

### CODE REVIEW

**Comment by tedteng** [comment]
This PR ready for review

File: pkg/controller/bastion/options.go [hunk 1]
```diff
@@ -163,30 +162,95 @@ func determineImageID(shoot *gardencorev1beta1.Shoot, providerConfig *awsv1alpha
...
-	for _, t := range types {
-		if strings.HasPrefix(t, "t") {
-			return t, nil
+	if len(tTypes.InstanceTypeOfferings) == 0 {
+		return "", fmt.Errorf("no t.* instance type available")
+	}
+
+	for _, t := range tTypes.InstanceTypeOfferings {
+		input := &ec2.DescribeInstanceTypesInput{
+			InstanceTypes: []*string{t.InstanceType},
```
Comment (petersutter) [inline pkg/controller/bastion/options.go:205]: why a for loop, why not pass tTypes to `InstanceTypes` as I suggested in https://github.com/gardener/gardener-extension-provider-aws/pull/700#issuecomment-1433229750?

File: pkg/controller/bastion/options.go [hunk 2]
```diff
@@ -163,30 +162,95 @@ func determineImageID(shoot *gardencorev1beta1.Shoot, providerConfig *awsv1alpha
...
+	if err != nil {
+		return "", err
 	}
 
-	// fallback to the first (hopefully smallest) other general purpose instance type
-	for _, t := range types {
-		if strings.HasPrefix(t, "t") {
-			return t, nil
+	if len(tTypes.InstanceTypeOfferings) == 0 {
+		return "", fmt.Errorf("no t.* instance type available")
```
Comment (petersutter) [inline pkg/controller/bastion/options.go:200]:
```suggestion
		return "", fmt.Errorf("no t* instance type offerings available")
```
this also to make the error distinguishable from

File: pkg/controller/bastion/options.go [hunk 3]
```diff
@@ -163,30 +162,95 @@ func determineImageID(shoot *gardencorev1beta1.Shoot, providerConfig *awsv1alpha
...
+func checkInstanceTypeOffering(ctx context.Context, filter string, awsClient *awsclient.Client) (*ec2.DescribeInstanceTypeOfferingsOutput, error) {
+	return awsClient.EC2.DescribeInstanceTypeOfferingsWithContext(ctx, &ec2.DescribeInstanceTypeOfferingsInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("instance-type"),
+				Values: []*string{aws.String(filter)},
+			},
+		},
+	})
+}
```
Comment (petersutter) [inline pkg/controller/bastion/options.go:256]: I do not see the point in extracting this one liner to a function, however if you want to keep it, I'd suggest to rename it to `getInstanceTypeOfferings`

File: pkg/controller/bastion/options.go [hunk 4]
```diff
@@ -163,30 +162,95 @@ func determineImageID(shoot *gardencorev1beta1.Shoot, providerConfig *awsv1alpha
...
+	imageArchitecture := imageInfo.Architecture
+
+	// default instance type
+	switch *imageArchitecture {
+	case "x86_64":
+		preferredType = "t2.nano"
+	case "arm64":
+		preferredType = "t4g.nano"
+	default:
+		return "", fmt.Errorf("image architecture not supported yet")
```
Comment (petersutter) [inline pkg/controller/bastion/options.go:181]:
```suggestion
		return "", fmt.Errorf("image architecture not supported")
```

File: pkg/controller/bastion/options.go (hunk 1 above)
Comment (tedteng) [reply pkg/controller/bastion/options.go]: committed a6b392709b95eb54d9638ad54de25af7997148ce

File: pkg/controller/bastion/options.go (hunk 2 above)
Comment (tedteng) [reply pkg/controller/bastion/options.go]: committed a6b392709b95eb54d9638ad54de25af7997148ce

File: pkg/controller/bastion/options.go (hunk 3 above)
Comment (tedteng) [reply pkg/controller/bastion/options.go]: committed a6b392709b95eb54d9638ad54de25af7997148ce

File: pkg/controller/bastion/options.go (hunk 4 above)
Comment (tedteng) [reply pkg/controller/bastion/options.go]: committed a6b392709b95eb54d9638ad54de25af7997148ce

File: pkg/controller/bastion/options.go [hunk 5]
```diff
@@ -163,30 +163,95 @@ func determineImageID(shoot *gardencorev1beta1.Shoot, providerConfig *awsv1alpha
...
+	result, err := awsClient.EC2.DescribeInstanceTypes(input)
+	if err != nil {
+		return "", err
+	}
+
+	if len(result.InstanceTypes) == 0 {
+		return "", fmt.Errorf("no t* instance type offerings available")
+	} else {
+		return *result.InstanceTypes[0].InstanceType, nil
+	}
```
Comment (petersutter) [inline pkg/controller/bastion/options.go:228]:
```suggestion
	result, err := awsClient.EC2.DescribeInstanceTypes(&ec2.DescribeInstanceTypesInput{
		InstanceTypes: aws.StringSlice(tTypeSet.UnsortedList()),
		Filters: []*ec2.Filter{
			{
				Name:   aws.String("processor-info.supported-architecture"),
				Values: []*string{imageArchitecture},
			},
		},
	})
	if err != nil {
		return "", err
	}

	if len(result.InstanceTypes) == 0 {
		return "", fmt.Errorf("no instance types returned for architecture %s and instance types list %v", *imageArchitecture, tTypeSet.UnsortedList())
	}

	return *result.InstanceTypes[0].InstanceType, nil
```

File: pkg/controller/bastion/options.go [hunk 6]
```diff
@@ -163,30 +163,95 @@ func determineImageID(shoot *gardencorev1beta1.Shoot, providerConfig *awsv1alpha
...
+	imageInfo, err := getImages(ctx, imageID, awsClient)
 	if err != nil {
-		return "", fmt.Errorf("failed to list instance types: %w", err)
+		return "", err
 	}
 
-	types := make([]string, len(offerings.InstanceTypeOfferings))
-	for i, offering := range offerings.InstanceTypeOfferings {
-		types[i] = *offering.InstanceType
+	imageArchitecture := imageInfo.Architecture
```
Comment (petersutter) [inline pkg/controller/bastion/options.go:173]: do we need to check if nil or is it guaranteed that `Architecture` is never nil?

File: pkg/controller/bastion/options.go (hunk 5 above)
Comment (tedteng) [reply pkg/controller/bastion/options.go]: committed  47b98430

File: pkg/controller/bastion/options.go (hunk 6 above)
Comment (tedteng) [reply pkg/controller/bastion/options.go:177]: committed  47b98430

File: pkg/controller/bastion/options.go [hunk 7]
```diff
@@ -163,30 +163,98 @@ func determineImageID(shoot *gardencorev1beta1.Shoot, providerConfig *awsv1alpha
...
+
+	if err != nil {
+		return "", err
+	}
+
+	if len(result.InstanceTypes) == 0 {
+		return "", fmt.Errorf("no instance types returned for architecture %s and instance types list %v", *imageArchitecture, tTypeSet.UnsortedList())
+	}
+
+	return *result.InstanceTypes[0].InstanceType, nil
```
Comment (petersutter) [inline pkg/controller/bastion/options.go:231]: same here, is `InstanceType` guaranteed to be not nil? otherwise this PR looks good to me

Comment (tedteng) [reply pkg/controller/bastion/options.go:235]: b93e187bfa57280390f5dcbc9f2fc1d9522b5529

**Review by petersutter — APPROVED** [review]

**Review by kon-angelo — APPROVED** [review]

**Pull request merged**